### PR TITLE
[FLINK-19333][python] Introduce BatchArrowPythonOverWindowAggregateFunctionOperator

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will introduce BatchArrowPythonOverWindowAggregateFunctionOperator to support Batch Pandas Over Window Aggregation*

## Brief change log

  - *Add BatchArrowPythonOverWindowAggregateFunctionOperator*

## Verifying this change

  - *BatchArrowPythonOverWindowAggregateFunctionOperatorTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+/**
+ * The Batch Arrow Python {@link AggregateFunction} Operator for Over Window Aggregation.
+ */
+@Internal
+public class BatchArrowPythonOverWindowAggregateFunctionOperator
+	extends AbstractBatchArrowPythonAggregateFunctionOperator {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final String SCHEMA_OVER_WINDOW_ARROW_CODER_URN = "flink:coder:schema:over_window_arrow:v1";
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java:47]:
```suggestion
	private static final String SCHEMA_OVER_WINDOW_ARROW_CODER_URN = "flink:coder:schema:batch_over_window:arrow:v1";
```

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	 */
+	private final long[] upperBoundary;
+
+	/**
+	 * Whether the specified position window is a range window.
+	 */
+	private final boolean[] isRangeWindows;
+
+	/**
+	 * The window index of the specified aggregate function belongs to.
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java:72]:
```suggestion
	 * The window index of the specified aggregate function belonging to.
```

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+	/**
+	 * The row time index of the input data.
+	 */
+	private final int inputTimeFieldIndex;
+
+	/**
+	 * The order of row time. True for ascending.
+	 */
+	private final boolean order;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java:84]:
```suggestion
	private final boolean asc;
```

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+		}
+		RowData forwardedFields = forwardedInputSerializer.copy(input);
+		forwardedInputQueue.add(forwardedFields);
+	}
+
+	@Override
+	protected void invokeCurrentBatch() throws Exception {
+		if (currentBatchCount > 0) {
+			arrowSerializer.finishCurrentBatch();
+			RowData[] boundaryInputData =
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java:177]: local variable boundaryInputData is not necessary

File: flink-python/pyflink/proto/flink-fn-execution.proto [hunk 5]
```diff
@@ -44,12 +44,33 @@ message UserDefinedFunction {
...
+    RANGE_UNBOUNDED_PRECEDING = 1;
+    RANGE_UNBOUNDED_FOLLOWING = 2;
+    RANGE_SLIDING = 3;
+    ROW_UNBOUNDED = 4;
+    ROW_UNBOUNDED_PRECEDING = 5;
+    ROW_UNBOUNDED_FOLLOWING = 6;
+    ROW_SLIDING = 7;
+  }
+  WindowType windowType = 1;
+  int64 lowerBoundary = 2;
```
Comment (dianfu) [inline flink-python/pyflink/proto/flink-fn-execution.proto:72]: lowerBoundary -> lower_boundary? Same for the other fields. This is to keep consistent for the code style of the proto file.

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+	@Override
+	public String getInputOutputCoderUrn() {
+		return SCHEMA_OVER_WINDOW_ARROW_CODER_URN;
+	}
+
+	private boolean isInCurrentOverWindow(RowData data, long time, boolean includeEqual) {
+		long dataTime = data.getTimestamp(inputTimeFieldIndex, 3).getMillisecond();
+		long diff = time - dataTime;
+		return diff > 0 && order || diff == 0 && includeEqual;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/batch/BatchArrowPythonOverWindowAggregateFunctionOperator.java:327]:
```suggestion
		return (diff > 0 && order) || (diff == 0 && includeEqual);
```
It makes the code more readable.

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks a lot for the PR. Good work!

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the update. I have addressed the comments at the latest commit.

**Review by dianfu — APPROVED** [review]
@HuangXingBo Thanks for the update. LGTM.

**Pull request closed without merging**